tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockClusterableServerFactory.cs

```diff
@@ -74,6 +79,45 @@ public IClusterableServer CreateServer(ClusterId clusterId, IClusterClock cluste
                             {
                                 _servers[mockMonitor.Object.Description.EndPoint].HasBeenRemoved = true;
                             });
+                    Action<string, TopologyVersion?> mockMonitorInvalidate = (reason, responseTopologyVersion) =>
+                    {
+                        var baseDescription = description;
+                        var oldDescription = mockMonitor.Object.Description;
+                        var newDescription = baseDescription.With(
+                            $"InvalidatedBecause:{reason}",
+                             lastUpdateTimestamp: DateTime.UtcNow,
+                             topologyVersion: responseTopologyVersion);
+                        mockMonitor.SetupGet(m => m.Description).Returns(newDescription);
+                        var serverDescriptionChangedEventArgs = new ServerDescriptionChangedEventArgs(
+                            oldDescription,
+                            newDescription);
+                        mockMonitor.Raise(m => m.DescriptionChanged += null, serverDescriptionChangedEventArgs);
+                    };
+                    mockMonitor
+                        .Setup(m => m.Invalidate(It.IsAny<string>(), It.IsAny<TopologyVersion?>()))
+                        .Callback(mockMonitorInvalidate);
+                    var mockConnection = new Mock<IConnectionHandle>();
+                    var mockConnectionPool = new Mock<IConnectionPool>();
+                    var poolGeneration = 0;
+                    var connectionGeneration = 0;
+                    // need to use a func to close over connectionGeneration
+                    mockConnection.Setup(c => c.Generation).Returns(valueFunction: () => connectionGeneration);
+                    // need to use a func to close over poolGeneration
+                    mockConnectionPool.Setup(p => p.Generation).Returns(valueFunction: () => poolGeneration);
+                    Action acquireConnectionCallback = () => { connectionGeneration = poolGeneration; };
+                    mockConnectionPool
+                        .Setup(p => p.AcquireConnection(It.IsAny<CancellationToken>()))
+                        .Callback(acquireConnectionCallback)
+                        .Returns(mockConnection.Object);
+                    mockConnectionPool
+                        .Setup(p => p.AcquireConnectionAsync(It.IsAny<CancellationToken>()))
+                        .Callback(acquireConnectionCallback)
+                        .ReturnsAsync(mockConnection.Object);
+                    mockConnectionPool.Setup(p => p.Clear()).Callback(() => { ++poolGeneration; });
+                    var mockConnectionPoolFactory = new Mock<IConnectionPoolFactory> { DefaultValue = DefaultValue.Mock };
+                    mockConnectionPoolFactory
+                        .Setup(f => f.CreateConnectionPool(It.IsAny<ServerId>(), endPoint))
+                        .Returns(mockConnectionPool.Object);
```

Comment:
The key thing here is the mocking up of the monitor's `Invalidate` function, which needs update the monitor description, notify the Server, and increment the mock connection pool's generation. We also ensure that the mock connection pool's `Acquire` functions  returns a mock connection with the correct generation number (i.e.  one that inherits the connection pool's generation number on acquisition and doesn't increment when the connection pool's generation number increments)